cli/src/commands/git/push.rs

```diff
@@ -396,46 +406,33 @@ fn print_commits_ready_to_push(
-                // Without color, it's hard to see at a glance if one bookmark
-                // among many was moved sideways (say). TODO: People on Discord
-                // suggest "Move bookmark ... forward by n commits",
-                // possibly "Move bookmark ... sideways (X forward, Y back)".
-                let msg = match to_direction(old_target, new_target) {
-                    BookmarkMoveDirection::Forward => {
-                        format!("Move forward bookmark {bookmark_name} from {old} to {new}")
-                    }
-                    BookmarkMoveDirection::Backward => {
-                        format!("Move backward bookmark {bookmark_name} from {old} to {new}")
-                    }
-                    BookmarkMoveDirection::Sideways => {
-                        format!("Move sideways bookmark {bookmark_name} from {old} to {new}")
-                    }
-                };
-                writeln!(formatter, "  {msg}")?;
-            }
-            (Some(old_target), None) => {
-                writeln!(
-                    formatter,
-                    "  Delete bookmark {bookmark_name} from {}",
-                    short_commit_hash(old_target)
-                )?;
-            }
-            (None, Some(new_target)) => {
-                writeln!(
-                    formatter,
-                    "  Add bookmark {bookmark_name} to {}",
-                    short_commit_hash(new_target)
-                )?;
-            }
+            // TODO: People on Discord
+            // suggest "Move bookmark ... forward by n commits",
+            // possibly "Move bookmark ... sideways (X forward, Y back)".
+            (Some(_), Some(_)) => write!(formatter, "  Move bookmark ",)?,
+            (Some(_), None) => write!(formatter, "  Delete bookmark ",)?,
+            (None, Some(_)) => write!(formatter, "  Add bookmark ",)?,
             (None, None) => {
                 panic!("Not pushing any change to bookmark {bookmark_name}");
             }
         }
+        write!(formatter.labeled("bookmarks"), "{bookmark_name}")?;
+        if let (Some(old_target), Some(new_target)) = (&update.old_target, &update.new_target) {
+            match to_direction(old_target, new_target) {
+                BookmarkMoveDirection::Forward => write!(formatter, " forward")?,
+                BookmarkMoveDirection::Backward => write!(formatter, " backward")?,
+                BookmarkMoveDirection::Sideways => write!(formatter, " sideways")?,
+            }
+        }
+        if let Some(old_target) = &update.old_target {
+            write!(formatter, " from ")?;
+            format_commit_shortest_prefix(formatter, repo, id_prefix_context, old_target)?;
+        }
+        if let Some(new_target) = &update.new_target {
+            write!(formatter, " to ")?;
+            format_commit_shortest_prefix(formatter, repo, id_prefix_context, new_target)?;
+        }
+        writeln!(formatter)?;
```

Comment:
Perhaps, this function (and `print_git_import_stats()`) can be extracted to templates similar to `bookmark_list`.

Each `(String, BookmarkPushUpdate)` could be encoded as `commit_templater::RefName`, but it's probably better to add `RefDiff` or `RefChange` template type. In `print_git_import_stats()`, we'll need to represent a diff of remote-ref to remote-ref.

Reply:
Yes they ought to be all templated I noticed that after I was done with this current version (hence the draft).